subsys/bluetooth/controller/ll_sw/ull_iso.c

```diff
@@ -625,24 +872,349 @@ uint8_t ll_read_iso_link_quality(uint16_t  handle,
+			/* Randomize the length [4..p_max_sdu] */
+			lll_rand_get(&rand_8, sizeof(rand_8));
+			rand_max_sdu = rand_8 * (cis->p_max_sdu - 4);
+			remaining_tx = 4 + (rand_max_sdu >> 8);
+			break;
+
+		case BT_HCI_ISO_TEST_MAX_SIZE_SDU:
+			LL_ASSERT(cis->p_max_sdu > 4);
+			remaining_tx = cis->p_max_sdu;
+			break;
+
+		default:
+			LL_ASSERT(0);
+			return;
+		}
+
+		if (remaining_tx > ISO_TEST_TX_BUFFER_SIZE) {
+			sdu.sdu_state = BT_ISO_START;
+		} else {
+			sdu.sdu_state = BT_ISO_SINGLE;
+		}
+
+		/* Configure SDU similarly to one delivered via HCI */
+		sdu.dbuf = tx_buffer;
+		sdu.cig_ref_point = cig->cig_ref_point;
+		sdu.target_event = cis->lll.event_count + (cis->lll.tx.flush_timeout > 1 ? 0 : 1);
+		sdu.iso_sdu_length = remaining_tx;
+
+		/* Send all SDU fragments */
+		do {
+			sdu.time_stamp = HAL_TICKER_TICKS_TO_US(ticks_at_expire);
+			sdu.size = MIN(remaining_tx, ISO_TEST_TX_BUFFER_SIZE);
+			memset(tx_buffer, 0, sdu.size);
+
+			/* If this is the first fragment of a framed SDU, inject the SDU
+			 * counter.
+			 */
+			if (sdu.size >= 4 && ((sdu.sdu_state == BT_ISO_START) ||
+					      (sdu.sdu_state == BT_ISO_SINGLE))) {
+				if (cis->framed) {
+					sdu_counter = (uint32_t)cis->hdr.test_mode.tx_sdu_counter;
+				} else {
+					/* Unframed. Get the next payload counter.
+					 *
+					 * BT 5.3, Vol 6, Part B, Section 7.1:
+					 * When using unframed PDUs, the SDU counter shall be equal
+					 * to the payload counter.
+					 */
+					source = isoal_source_get(source_handle);
+					pdu_production = &source->pdu_production;
+
+					sdu_counter = MAX(pdu_production->payload_number,
+							  (sdu.target_event *
+							   cis->lll.tx.burst_number));
+				}
+
+				sys_put_le32(sdu_counter, tx_buffer);
+			}
+
+			/* Send to ISOAL */
+			err = isoal_tx_sdu_fragment(source_handle, &sdu);
+			LL_ASSERT(!err);
+
+			remaining_tx -= sdu.size;
+
+			if (remaining_tx > ISO_TEST_TX_BUFFER_SIZE) {
+				sdu.sdu_state = BT_ISO_CONT;
+			} else {
+				sdu.sdu_state = BT_ISO_END;
+			}
+		} while (remaining_tx);
+
+		cis->hdr.test_mode.tx_sdu_counter++;
+
+	} else if (IS_ADV_ISO_HANDLE(handle)) {
+		/* FIXME: Implement for broadcaster */
+	} else {
+		LL_ASSERT(0);
+	}
+}
+
+/* An ISO interval has elapsed for a Connected Isochronous Group */
+void ll_iso_transmit_test_cig_interval(uint16_t handle, uint32_t ticks_at_expire)
+{
+	struct ll_conn_iso_stream *cis;
+	struct ll_conn_iso_group *cig;
+	uint32_t sdu_interval;
+	uint32_t iso_interval;
+	uint16_t handle_iter;
+	uint64_t sdu_counter;
+	uint8_t tx_sdu_count;
+
+	cig = ll_conn_iso_group_get(handle);
+	LL_ASSERT(cig);
+
+	handle_iter = UINT16_MAX;
+
+	if (cig->lll.role) {
+		/* Peripheral */
+		sdu_interval = cig->p_sdu_interval;
+	} else {
+		/* Central */
+		sdu_interval = cig->c_sdu_interval;
+	}
+
+	iso_interval = cig->iso_interval * PERIODIC_INT_UNIT_US;
+
+	/* Handle ISO Transmit Test for all active CISes in the group */
+	for (int i = 0; i < cig->lll.num_cis; i++)  {
```

Comment:
Suggest to use same type for the iterator and the value it's compared against
```suggestion
	for (uint8_t i = 0; i < cig->lll.num_cis; i++)  {
```